WIP | Improve type parameter suggestion heuristic for missing types

Kivooeo opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/140073)*
<!-- homu-ignore:end -->

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? @jieyouxu 
-->
<!-- homu-ignore:end -->

I consider this solution a starting point and not final. I'm open to your ideas and suggestions for further improvements to the heuristic or implementation.

### Summary

This PR improves the heuristic used by `rustc` to suggest type parameters for missing types, addressing issue rust-lang/rust#139999. The issue occurs when `rustc` mistakenly suggests adding a type parameter for a missing type (e.g., `ForgotToImport`) that is likely a concrete type rather than a generic. This can lead to confusing error messages, as shown in the example:

```rust
pub fn do_stuff() -> Option<ForgotToImport> {
    None
}
```

Current output suggests adding a type parameter:

```
help: you might be missing a type parameter
  |
1 | pub fn do_stuff<ForgotToImport>() -> Option<ForgotToImport> {
  |                ++++++++++++++++
```

This PR refines the heuristic to avoid such suggestions for names that are likely concrete types.

### Changes

The fix modifies the logic in the type suggestion code to skip type parameter suggestions when the missing type name:

1. Is not a common generic name (e.g., `Item`, `Output`, `Error`).
2. Uses UpperCamelCase, which is typical for concrete types.
3. Is longer than 3 characters, as type parameters are often short (e.g., `T`, `U`).

### Example

With this change, the example code above will no longer suggest a type parameter for `ForgotToImport`, resulting in a cleaner error message:

```
error[E0412]: cannot find type `ForgotToImport` in this scope
 --> src/lib.rs:1:29
  |
1 | pub fn do_stuff() -> Option<ForgotToImport> {
  |                             ^^^^^^^^^^^^^^ not found in this scope
```

### Rationale

This approach balances precision and simplicity:

- UpperCamelCase is a strong indicator of concrete types in Rust, as type parameters are often single letters or specific generic names.
- The length check (&gt;3) reduces false positives for longer names, which are less common for generics.
- Preserving suggestions for well-known generic names ensures backward compatibility and avoids breaking valid cases.

### Related Issue

fixes rust-lang/rust#139999

cc @Wilfred (issue reporter)

Kivooeo committed:
fix for #139999

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14558958625/job/40839694632) [(plain)](https://github.com/rust-lang/rust/commit/037c08ef9eb8c34201a7eed46d691e66ee3ebe46/checks/40839694632/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.891 Building wheels for collected packages: reuse
#12 2.892   Building wheel for reuse (pyproject.toml): started
#12 3.106   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.107   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.107   Stored in directory: /tmp/pip-ephem-wheel-cache-el0scsh5/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.110 Successfully built reuse
#12 3.110 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.523 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.524 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.065 Collecting virtualenv
#12 4.119   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.310      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 22.9 MB/s eta 0:00:00
#12 4.370 Collecting platformdirs<5,>=3.9.1
#12 4.378   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.419 Collecting filelock<4,>=3.12.2
#12 4.426   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.447 Collecting distlib<1,>=0.3.7
#12 4.454   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.468      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 38.8 MB/s eta 0:00:00
#12 4.550 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.733 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 4.734 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.8s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      139200 kB
DirectMap2M:     7200768 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.228
---
[TIMING] core::build_steps::tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 5984 files
tidy check
tidy error: file `tests/ui/traits/issue-139999.rs` must begin with a descriptive name, consider `{reason}-issue-139999.rs`
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:48
  local time: Sun Apr 20 11:29:19 UTC 2025
  network time: Sun, 20 Apr 2025 11:29:19 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

oli-obk commented on tests/ui/traits/missing-type-param-suggestion-issue-139999.rs:
Also test that a common name like `Item` still suggest making it generic

oli-obk commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -2772,6 +2772,32 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
+                        && sugg.chars().skip(1).any(|c| c.is_lowercase());
+
+                    // If it's not a known generic name and looks like a concrete type, skip the suggestion.
+                    if !is_common_generic && looks_like_camel_case && sugg.len() > 3 {
```
the len 3 check is always triggered on const parameter suggestions. check the length of `ident` instead. Also add a test for const generics, too

Kivooeo replied:
can you please show me example of test for const, because i dont see how can i do it
```rust
pub fn do_stuff<const T: usize>() -> Option<T> {
    None
}
``` 
this surely wrong because trigger not this error
and 
```rust
pub fn do_stuff<const T: usize>() -> Option<ForgetType> {
    None
}
```
this works fine

and i cant do something like this 
```rust
pub fn do_stuff() -> Option<const T: usize> {
    None
}
```

so i need an example of how test it correctly

fmease replied:
```rust
fn f() -> [(); N] {}
```

Kivooeo replied:
this one works just fine i guess?
```rust
error[E0425]: cannot find value `N` in this scope
 --> C:\Users\Love\projects\asfgas\src/main.rs:1:16
  |
1 | fn f() -> [(); N] {}
  |                ^ not found in this scope
  |
help: you might be missing a const parameter
  |
1 | fn f<const N: /* Type */>() -> [(); N] {}
  |     +++++++++++++++++++++
```

fmease replied:
Oli is saying that for const parameters you currently perform checks like `common_generic_names.contains("const N: usize")` which will *never* fail! You should compare against the original ident (in this case `N`)

fmease replied:
Note however that for const parameters, the list `common_generic_names` doesn't make sense, it's specifically for type parameters, so you should be comparing against a different list or none at all

Kivooeo replied:
yes i got it, but i was wondering how do i write tests for this const case more correctly, and i also want to ask if name is full uppercase, should we consider it as generic?

fmease replied:
We should either use a different list for const params (of common const parameter names of length>1 (idk which ones could be considered common (`LEN`? `LENGTH`?, `SIZE`? perhaps I dunno)) or don't do such a check for them at all

fmease replied:
> and i also want to ask if name is full uppercase, should we consider it as generic?

I guess not necessarily, similar to (generic) type parameters. Since `CONST` can just as well refer to a const item. E.g.:

```rs
const ELEMENTS: u32 = 3;
fn f() -> [i32; ELEMENTS] { todo!() }
```

oli-obk commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -2772,6 +2772,32 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
+                    let looks_like_camel_case = sugg
+                        .chars()
+                        .next()
+                        .map(|c| c.is_uppercase())
+                        .unwrap_or(false)
```
```suggestion
                        .is_some_and(|c| c.is_uppercase())
```

oli-obk commented:
To close the issue when this PR is merged, change your main comment to contain `fixes #123456` (also please remove that from the title

fmease commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -2772,6 +2772,32 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
+                    //
+                    // This approach may produce false positives or negatives, but works well in most cases.
+
+                    let common_generic_names: &[&str] = &[
```
```suggestion
                    let common_type_param_names: &[&str] = &[
```

since it doesn't make sense for const params

fmease replied:
Also, I'd like to draw attention to the

```rust
    && segment.ident.name != kw::SelfUpper
    && segment.ident.name != kw::Dyn =>
```

at the very start of the function. We're already filtering out certain identifiers (in this case `Self` and `dyn`). I'd like us to unify this check and the `common_generic_names` one is some way or another

Kivooeo replied:
this is good point, will it be better to replace my logic right after this `let (ident, span) = ...` or integrate them inside the `match`?

mejrs replied:
I think what @fmease might expect is a change like...

```rust

        let [segment] = segments else {
               return None;
         };

          // comment here explaining why we skip these
         let skip_names = [kw::SelfUpper, kw::Dyn, sym::This, sym::Error, /* etc */];

         if !segment.has_generic_args && !skip_names.contains(segment.ident.name) {
                 return None;
         }
         let ident = segment.ident.to_string();
         let span =  segment.ident.span;
```

I personally much prefer let_else over a match in this case. Feel free to write it with a match if you prefer.

Also, not all these names are available as symbols in `sym`. Feel free to add them there as needed,

Kivooeo commented on compiler/rustc_resolve/src/late/diagnostics.rs:
```diff
@@ -2764,14 +2764,39 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
                     let (msg, sugg) = match source {
                         PathSource::Type | PathSource::PreciseCapturingArg(TypeNS) => {
-                            ("you might be missing a type parameter", ident)
+                            ("you might be missing a type parameter", ident.clone())
```
would love to hear ideas about how to remove this clone

Kivooeo commented:
@oli-obk @fmease check this changes, replaced `sugg` with `ident` and added more test cases.
i played around this check at function start and i didnt liked it much because split condition leads to not all cases are covered, and moving entire logic on top is worse because the place it current located filtering out `main` which is good for more precise error messaging.

and i also didnt liked idea about renaming array of common names, because it's just common generic names that looks like UpperCamelCase and it's len > 3, so this array and check is needed to remove false positive results

alex-semenyuk commented:
@Kivooeo
Thanks for your contribution.
Form wg-triage. Any updates on this PR?

Kivooeo commented:
It's basically waiting for review

alex-semenyuk commented:
@rustbot remove-label S-waiting-on-author add-label S-waiting-on-review

Kivooeo commented:
@fmease given to gap of time, we need to revise rules, which we we'll use to use heuristics and filter generics out of forgetted things, 

I also wonder if we even need such mechanism but let's say for start thay yes

Then here me proposing a start version of rules and then work on it before implementation itself (which wont be a problem)

1. less than eight length
2. check camelcase
3. we might want somehow check scope context and signature if it have any generics for sure
4. id keep a list with common names
5. also we can module some kind of score system

wesleywiser commented:
Hi @oli-obk and @fmease, it looks like the author has some questions [here](https://github.com/rust-lang/rust/pull/140073#issuecomment-3172076273) that would be helpful in getting this ready for another review.

bors commented:
:umbrella: The latest upstream changes (presumably #146499) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

fmease reviewed (changes requested):
I'm sorry for the delay.

> The fix modifies the logic in the type suggestion code to skip type parameter suggestions when the missing type name:
>
> 1. Is not a common generic name (e.g., `Item`, `Output`, `Error`).
> 2. Uses UpperCamelCase, which is typical for concrete types.
> 3. Is longer than 3 characters, as type parameters are often short (e.g., `T`, `U`).

I guess in practice it's true that these are unlikely meant to be type parameters. It does kind of punish users who make an effort in descriptively naming their type/const params.

I disagree with some of these "arbitrarily chosen" exceptions like `Error` (I have named ty params that before over just `E`). Strictly speaking, upper camel case *is* the naming convention of all non-primitive types which includes type params.

In order to unblock this, could you please remove the list of exceptions together with the upper camel case heuristic and only keep the codepoint count check.

I was about to say let's raise the limit from 3 to 5 but actually let's make this heuristic even simpler and *require* the ident to be a single upper-case char (which we've already computed). On master, that ident only needs to be one if we're outside of a generic argument list.

It goes directly against my previous statement ("punishes" people with descriptive generic param names) but I find it more important that the heuristic is a simple as possible (without greatly sacrificing the diagnostic quality). If you believe count==1 regresses to many `tests/ui/`s, then go with my suggest count==5 (still no skip list or casing detection please).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/140073/6b3ae3f6e45a33c2d95fa0362c9b2593e567fd34..e94d1923ea6dfecdeb0daa48d32a115c502d9ea6)*

Kivooeo commented:
closing since i think now such heuristic isn't good and not worth fixing
